File: scripts/train-intent-ranker.mjs
/**
 * Intent ranker trainer (goal doc Fáze 2) — offline development tooling.
 *
 * Pure Node, zero npm dependencies:
 *  1. loads scripts/data/intent-ranker-dataset.json (features.v1 vectors +
 *     heuristic teacher scores),
 *  2. trains a tiny pairwise MLP (RankNet-style logistic loss on
 *     better/worse pairs; heuristic score is the TEACHER signal, never proof
 *     of musical quality),
 *  3. serializes the trained weights as a minimal ONNX graph
 *     (Gemm+Relu ×3 → Gemm; input [N, featureCount] → output [N, 1]),
 *  4. writes public/models/intent-ranker-v1.onnx + manifest + a validation
 *     report (train/val pairwise accuracy vs the heuristic teacher, top-1
 *     agreement, Spearman rank correlation on held-out groups).
 *
 * Split is by GROUP (seed × style) — no leakage between train and val.
 *
 * Run: node scripts/train-intent-ranker.mjs
 */
import { readFileSync, writeFileSync, mkdirSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");
const datasetPath = path.join(ROOT, "scripts", "data", "intent-ranker-dataset.json");
const modelsDir = path.join(ROOT, "public", "models");

const dataset = JSON.parse(readFileSync(datasetPath, "utf8"));
const featureNames = dataset.featureNames ?? null;
const HIDDEN = [64, 32, 16];
const EPOCHS = 150;
const LEARNING_RATE = 0.08;
const SEED = 0x5eed;

function mulberry32(seed) {
  let a = seed >>> 0;
  return () => {
    a = (a + 0x6d2b79f5) >>> 0;
    let t = Math.imul(a ^ (a >>> 15), 1 | a);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}
const rng = mulberry32(SEED);

// ── flatten dataset ──
const samples = [];
for (const group of dataset.groups) {
  const rows = group.candidates.map((candidate) => ({
    x: candidate.features,
    score: candidate.heuristicScore,
    index: candidate.index,
    groupKey: group.groupKey,
  }));
  samples.push(...rows);
}
const featureCount = samples[0]?.x.length ?? 0;
if (!featureCount) throw new Error("empty dataset");
if (featureNames && featureNames.length !== featureCount) throw new Error("feature count mismatch");

// deterministic group split: 80% train / 20% val by group hash
const groupKeys = [...new Set(samples.map((s) => s.groupKey))].sort();
const valKeys = new Set();
for (const [i, key] of groupKeys.entries()) if (i % 5 === 0) valKeys.add(key);
const train = samples.filter((s) => !valKeys.has(s.groupKey));
const val = samples.filter((s) => valKeys.has(s.groupKey));
console.log(`[train] samples total=${samples.length} train=${train.length} val=${val.length} features=${featureCount}`);

// ── tiny MLP forward/backward (RankNet pairwise logistic loss) ──
const layers = [];
let prevSize = featureCount;
for (const size of [...HIDDEN, 1]) {
  layers.push({
    w: Array.from({ length: size }, () => Array.from({ length: prevSize }, () => (rng() - 0.5) * 0.4)),
    b: new Array(size).fill(0),
    mW: new Array(size).fill(0).map(() => new Array(prevSize).fill(0)),
    mB: new Array(size).fill(0),
  });
  prevSize = size;
}

function forward(x, weights = layers) {
  const activations = [x];
  let current = x;
  for (let l = 0; l < weights.length; l++) {
    const layer = weights[l];
    const out = new Array(layer.w.length);
    const isLast = l === weights.length - 1;
    for (let j = 0; j < layer.w.length; j++) {
      let sum = layer.b[j];
      const row = layer.w[j];
      for (let k = 0; k < current.length; k++) sum += row[k] * current[k];
      out[j] = isLast ? sum : Math.max(0, sum); // ReLU hidden, linear output
    }
    activations.push(out);
    current = out;
  }
  return activations;
}

function trainPair(xBetter, xWorse, lr) {
  const actB = forward(xBetter);
  const actW = forward(xWorse);
  const diff = actB[actB.length - 1][0] - actW[actW.length - 1][0];
  // -log sigmoid(diff); dLoss/dDiff = -sigmoid(-diff) = -1/(1+e^diff)
  const grad = -1 / (1 + Math.exp(diff));
  // backprop both paths with ±grad on the output pre-activation
  for (const [activation, sign] of [
    [actB, grad],
    [actW, -grad],
  ]) {
    let delta = new Array(activation[activation.length - 1].length).fill(sign);
    for (let l = layers.length - 1; l >= 0; l--) {
      const layer = layers[l];
      const input = activation[l];
      const nextDelta = new Array(input.length).fill(0);
      for (let j = 0; j < layer.w.length; j++) {
        const g = delta[j];
        const row = layer.w[j];
        for (let k = 0; k < input.length; k++) {
          nextDelta[k] += row[k] * g;
          layer.mW[j][k] += g * input[k];
          // ReLU derivative on hidden layers only
          if (l > 0 && input[k] <= 0) nextDelta[k] = 0;
        }
        layer.mB[j] += g;
      }
      delta = nextDelta;
    }
  }
  return Math.log(1 + Math.exp(Math.min(30, diff)));
}

// build deterministic pairwise list
function pairsFor(rows) {
  const pairs = [];
  for (const groupKey of [...new Set(rows.map((r) => r.groupKey))].sort()) {
    const groupRows = rows.filter((r) => r.groupKey === groupKey);
    for (let a = 0; a < groupRows.length; a++) {
      for (let b = a + 1; b < groupRows.length; b++) {
        const better = groupRows[a].score >= groupRows[b].score ? groupRows[a] : groupRows[b];
        const worse = better === groupRows[a] ? groupRows[b] : groupRows[a];
        if (better.score !== worse.score) pairs.push([better.x, worse.x]);
      }
    }
  }
  return pairs.sort((p, q) => JSON.stringify(p[0]).localeCompare(JSON.stringify(q[0])));
}

const trainPairs = pairsFor(train);
const valPairs = pairsFor(val);
console.log(`[train] pairs train=${trainPairs.length} val=${valPairs.length}`);

for (let epoch = 0; epoch < EPOCHS; epoch++) {
  let loss = 0;
  const shuffled = [...trainPairs];
  const shuffleRng = mulberry32(SEED + epoch);
  for (let i = shuffled.length - 1; i > 0; i--) {
    const j = Math.floor(shuffleRng() * (i + 1));
    [shuffled[i], shuffled[j]] = [shuffled[j], shuffled[i]];
  }
  const lr = LEARNING_RATE;
  for (const [xB, xW] of shuffled) loss += trainPair(xB, xW, lr);
  // apply accumulated gradients
  const scale = (1 / shuffled.length) * lr;
  for (const layer of layers) {
    for (let j = 0; j < layer.w.length; j++) {
      for (let k = 0; k < layer.w[j].length; k++) {
        layer.w[j][k] -= scale * layer.mW[j][k];
        layer.mW[j][k] = 0;
      }
      layer.b[j] -= scale * layer.mB[j];
      layer.mB[j] = 0;
    }
  }
  if (epoch % 100 === 0 || epoch === EPOCHS - 1) console.log(`[train] epoch ${epoch} loss=${(loss / shuffled.length).toFixed(5)}`);
}

// ── validation: does the model AGREE with the teacher on held-out pairs? ──
function modelScore(x) {
  const activation = forward(x);
  return activation[activation.length - 1][0];
}
function pairwiseAccuracy(rows) {
  const pairs = pairsFor(rows);
  let agree = 0;
  for (const [xB, xW] of pairs) if (modelScore(xB) > modelScore(xW)) agree += 1;
  return pairs.length > 0 ? agree / pairs.length : 1;
}
function top1Agreement(rows) {
  const groupKeys = [...new Set(rows.map((r) => r.groupKey))].sort();
  let agree = 0;
  let total = 0;
  for (const groupKey of groupKeys) {
    const groupRows = rows.filter((r) => r.groupKey === groupKey);
    if (groupRows.length < 2) continue;
    const bestHeuristic = [...groupRows].sort((a, b) => b.score - a.score)[0];
    const bestModel = [...groupRows].sort((a, b) => modelScore(b.x) - modelScore(a.x))[0];
    if (bestHeuristic.index === bestModel.index) agree += 1;
    total += 1;
  }
  return total > 0 ? agree / total : 1;
}
function spearman(rows) {
  const groupKeys = [...new Set(rows.map((r) => r.groupKey))].sort();
  let sumRho = 0;
  let count = 0;
  for (const groupKey of groupKeys) {
    const groupRows = rows.filter((r) => r.groupKey === groupKey);
    if (groupRows.length < 3) continue;
    const byHeuristic = [...groupRows].sort((a, b) => b.score - a.score).map((r) => r.index);
    const byModel = [...groupRows].sort((a, b) => modelScore(b.x) - modelScore(a.x)).map((r) => r.index);
    const d2 = byHeuristic.reduce((sum, index, rank) => sum + (byModel.indexOf(index) - rank) ** 2, 0);
    const n = groupRows.length;
    sumRho += 1 - (6 * d2) / (n * (n * n - 1));
    count += 1;
  }
  return count > 0 ? sumRho / count : 1;
}

const report = {
  datasetVersion: dataset.datasetVersion,
  featureVersion: "features.v1",
  featureCount,
  hidden: HIDDEN,
  epochs: EPOCHS,
  trainPairs: trainPairs.length,
  valPairs: valPairs.length,
  valPairwiseAccuracyVsHeuristic: Number(pairwiseAccuracy(val).toFixed(4)),
  trainPairwiseAccuracyVsHeuristic: Number(pairwiseAccuracy(train).toFixed(4)),
  valTop1AgreementWithHeuristic: Number(top1Agreement(val).toFixed(4)),
  valSpearmanVsHeuristic: Number(spearman(val).toFixed(4)),
};
console.log("[train] report:", JSON.stringify(report, null, 2));

// ── minimal ONNX protobuf writer (float32, Gemm+Relu graph) ──
function varint(value) {
  const bytes = [];
  let v = value;
  while (v > 0x7f) {
    bytes.push((v & 0x7f) | 0x80);
    v >>>= 7;
  }
  bytes.push(v);
  return bytes;
}
function tag(field, wireType) {
  return varint((field << 3) | wireType);
}
function lenDelimited(field, payload) {
  return [...tag(field, 2), ...varint(payload.length), ...payload];
}
function strField(field, text) {
  return lenDelimited(field, [...Buffer.from(text, "utf8")]);
}
function floatField(field, value) {
  const bytes = [];
  new DataView(new ArrayBuffer(4)).constructor;
  const view = new DataView(new ArrayBuffer(4));
  view.setFloat32(0, value, true);
  for (let i = 0; i < 4; i++) bytes.push(view.getUint8(i));
  return [...tag(field, 5), ...bytes];
}
function int64Field(field, value) {
  return [...tag(field, 0), ...varint(value)];
}

function tensor(name, dims, data) {
  const out = [...strField(8, name)];
  for (const dim of dims) out.push(...tag(1, 0), ...varint(dim));
  out.push(...tag(2, 0), ...varint(1)); // data_type FLOAT
  out.push(...tag(9, 2), ...varint(data.length * 4));
  const view = new DataView(new ArrayBuffer(4));
  for (const f of data) {
    view.setFloat32(0, f, true);
    for (let i = 0; i < 4; i++) out.push(view.getUint8(i));
  }
  return out;
}
function node(opType, name, inputs, outputs, attrs = []) {
  let out = [];
  for (const input of inputs) out.push(...strField(1, input));
  for (const output of outputs) out.push(...strField(2, output));
  out.push(...strField(3, name));
  out.push(...strField(4, opType));
  // AttributeProto: name=1 (string), f=2 (float), i=3 (int64)
  for (const attr of attrs) out.push(...lenDelimited(5, attr));
  return lenDelimited(1, out);
}
function attrF(name, value) {
  return [...strField(1, name), ...floatField(2, value)];
}
function attrI(name, value) {
  return [...strField(1, name), ...int64Field(3, value)];
}
function valueInfoField(field, name, dims) {
  const shape = [];
  for (const dim of dims) shape.push(...lenDelimited(1, [...tag(1, 0), ...varint(dim)]));
  const tensorType = [...tag(1, 0), ...varint(1), ...lenDelimited(2, lenDelimited(1, shape))];
  return lenDelimited(field, [...strField(1, name), ...lenDelimited(2, lenDelimited(1, tensorType))]);
}

const graphName = "intent-ranker-v1";
const graphNodes = [];
const graphInitializers = [];
const dimsFor = (size) => [1, size];
let current = "features";
let index = 0;
for (let l = 0; l < layers.length; l++) {
  const size = layers[l].w.length;
  const flat = layers[l].w.flat();
  graphInitializers.push(tensor(`W${l}`, [size, layers[l].w[0].length], flat));
  graphInitializers.push(tensor(`B${l}`, [size], layers[l].b));
  const gemmOut = `gemm${l}`;
  graphNodes.push(
    node(
      "Gemm",
      `gemm_${l}`,
      [current, `W${l}`, `B${l}`],
      [gemmOut],
      [attrF("alpha", 1.0), attrF("beta", 1.0), attrI("transA", 0), attrI("transB", 1)],
    ),
  );
  current = gemmOut;
  if (l < layers.length - 1) {
    const reluOut = `relu${l}`;
    graphNodes.push(node("Relu", `relu_${l}`, [current], [reluOut]));
    current = reluOut;
  }
  index += 1;
}
void index;
const graph = [
  ...graphNodes,
  ...lenDelimited(2, [...strField(1, graphName)]),
  ...graphInitializers.flatMap((init) => lenDelimited(5, init)),
  ...valueInfoField(11, "features", dimsFor(featureCount)),
  ...valueInfoField(12, "score", dimsFor(1)),
];
const model = [
  ...int64Field(1, 8),
  ...strField(2, "pulse-forge-intent-trainer"),
  ...lenDelimited(7, graph),
  ...lenDelimited(8, [...strField(1, ""), ...int64Field(2, 13)]),
];

mkdirSync(modelsDir, { recursive: true });
const onnxPath = path.join(modelsDir, "intent-ranker-v1.onnx");
writeFileSync(onnxPath, Buffer.from(model));
const { createHash } = await import("node:crypto");
const modelHash = createHash("sha256").update(Buffer.from(model)).digest("hex");
console.log(`[train] model written ${onnxPath} (${(model.length / 1024).toFixed(1)} kB) sha256=${modelHash.slice(0, 16)}…`);

const manifest = {
  rankerVersion: "ranker.v1",
  featureVersion: "features.v1",
  normalizationId: "norm.fixed.v1",
  featureCount,
  modelPath: "/models/intent-ranker-v1.onnx",
  inputName: "features",
  outputName: "score",
  modelHash,
  hidden: HIDDEN,
  report,
};
writeFileSync(path.join(modelsDir, "intent-ranker-v1.manifest.json"), JSON.stringify(manifest, null, 2));
writeFileSync(path.join(ROOT, "scripts", "data", "intent-ranker-validation.json"), JSON.stringify(report, null, 2));
console.log("[train] manifest written");

File: scripts/train-intent-ranker.py
"""Intent ranker trainer (goal doc Fáze 2) — offline development tooling.

Pure numpy + onnx: no torch, no network, no browser involvement.

1. Loads scripts/data/intent-ranker-dataset.json (features.v1 vectors +
   heuristic teacher scores).
2. Trains a tiny pairwise MLP (RankNet-style logistic loss on better/worse
   pairs; the heuristic score is the TEACHER signal, never proof of musical
   quality) with plain Adam on numpy.
3. Exports the trained weights as a minimal ONNX graph
   (Gemm+Relu x3 -> Gemm; input [N, featureCount] -> output [N, 1]) via the
   official onnx helper, validated by onnx.checker AND onnxruntime inference.
4. Writes public/models/intent-ranker-v1.onnx, the manifest and a validation
   report (train/val pairwise accuracy vs the heuristic teacher, top-1
   agreement, Spearman rank correlation on held-out groups).

Split is by GROUP (seed x style) — no leakage between train and val.
Run: python scripts/train-intent-ranker.py
"""
from __future__ import annotations

import hashlib
import json
import path_helper  # noqa: F401  (adds repo root when run from anywhere)
from pathlib import Path

import numpy as np
import onnx
from onnx import TensorProto, checker, helper, numpy_helper

ROOT = Path(__file__).resolve().parent.parent
DATASET_PATH = ROOT / "scripts" / "data" / "intent-ranker-dataset.json"
MODELS_DIR = ROOT / "public" / "models"

HIDDEN = [64, 32, 16]
EPOCHS = 1200
LR = 0.001
SEED = 0x5EED


def sigmoid(value: float) -> float:
    value = max(-30.0, min(30.0, value))
    return 1.0 / (1.0 + np.exp(-value))


def make_rng(seed: int):
    return np.random.default_rng(seed)


class MLP:
    """Tiny ReLU MLP with manual Adam on a pairwise logistic loss."""

    def __init__(self, sizes: list[int], rng) -> None:
        self.weights: list[np.ndarray] = []
        self.biases: list[np.ndarray] = []
        for index in range(1, len(sizes)):
            scale = np.sqrt(2.0 / sizes[index - 1])
            self.weights.append(rng.normal(0, scale, (sizes[index - 1], sizes[index])).astype(np.float64))
            self.biases.append(np.zeros(sizes[index], dtype=np.float64))
        # Adam moments
        self.mW = [np.zeros_like(w) for w in self.weights]
        self.vW = [np.zeros_like(w) for w in self.weights]
        self.mB = [np.zeros_like(b) for b in self.biases]
        self.vB = [np.zeros_like(b) for b in self.biases]

    def forward(self, x: np.ndarray) -> list[np.ndarray]:
        activations = [x]
        last = len(self.weights) - 1
        for index, (weight, bias) in enumerate(zip(self.weights, self.biases)):
            out = activations[-1] @ weight + bias
            if index != last:
                out = np.maximum(0.0, out)  # ReLU hidden
            activations.append(out)
        return activations

    def pair_gradients(self, x_better: np.ndarray, x_worse: np.ndarray) -> tuple[float, list, list]:
        act_better = self.forward(x_better)
        act_worse = self.forward(x_worse)
        diff = float(act_better[-1].sum() - act_worse[-1].sum())
        loss = np.log1p(np.exp(np.clip(-diff, -30, 30)))  # -log sigmoid(diff)
        grad_output = -sigmoid(-diff)  # d(-log sigmoid(diff))/d(diff)

        # Accumulate BOTH paths' gradients — per-path updates would run two
        # opposing steps per pair and invert the learned ordering.
        total_gW = [np.zeros_like(w) for w in self.weights]
        total_gB = [np.zeros_like(b) for b in self.biases]
        for activation, sign in ((act_better, grad_output), (act_worse, -grad_output)):
            delta = np.full(activation[-1].shape, sign, dtype=np.float64)
            for index in range(len(self.weights) - 1, -1, -1):
                input_activation = activation[index]
                total_gW[index] += np.outer(input_activation, delta)
                total_gB[index] += delta.sum(axis=0)
                delta = delta @ self.weights[index].T
                if index > 0:
                    delta[input_activation <= 0] = 0  # ReLU derivative
        return float(loss), total_gW, total_gB

    def score(self, x: np.ndarray) -> np.ndarray:
        return self.forward(x)[-1]

    def export_onnx(self, feature_count: int) -> bytes:
        initializers = []
        nodes = []
        current = "features"
        for index, (weight, bias) in enumerate(zip(self.weights, self.biases)):
            weight_name, bias_name = f"W{index}", f"B{index}"
            initializers.append(numpy_helper.from_array(weight.astype(np.float32), weight_name))
            initializers.append(numpy_helper.from_array(bias.astype(np.float32), bias_name))
            gemm_out = f"gemm{index}"
            nodes.append(
                helper.make_node(
                    "Gemm",
                    [current, weight_name, bias_name],
                    [gemm_out],
                    name=f"gemm_{index}",
                    alpha=1.0,
                    beta=1.0,
                    transA=0,
                    # We store each weight as [input, output], matching the
                    # numpy forward pass, so Gemm must keep B untransposed.
                    transB=0,
                )
            )
            current = gemm_out
            if index < len(self.weights) - 1:
                relu_out = f"relu{index}"
                nodes.append(helper.make_node("Relu", [current], [relu_out], name=f"relu_{index}"))
                current = relu_out

        # Keep the public output name separate from the internal Gemm tensor.
        # This makes the graph contract explicit and avoids checker/runtime
        # ambiguity when the final Gemm is optimized or renamed.
        nodes.append(helper.make_node("Identity", [current], ["score"], name="score_output"))

        input_tensor = helper.make_tensor_value_info("features", TensorProto.FLOAT, [None, feature_count])
        output_tensor = helper.make_tensor_value_info("score", TensorProto.FLOAT, [None, 1])
        graph = helper.make_graph(
            nodes,
            "intent-ranker-v1",
            [input_tensor],
            [output_tensor],
            initializer=initializers,
        )
        model = helper.make_model(
            graph,
            producer_name="pulse-forge-intent-trainer",
            opset_imports=[helper.make_opsetid("", 13)],
        )
        model.ir_version = 8
        checker.check_model(model)
        return model.SerializeToString()


def pairs_for(rows: list[dict], rng) -> list[tuple[dict, dict]]:
    pairs: list[tuple[dict, dict]] = []
    group_keys = sorted({row["groupKey"] for row in rows})
    for group_key in group_keys:
        group_rows = [row for row in rows if row["groupKey"] == group_key]
        for a in range(len(group_rows)):
            for b in range(a + 1, len(group_rows)):
                better, worse = (
                    (group_rows[a], group_rows[b])
                    if group_rows[a]["score"] >= group_rows[b]["score"]
                    else (group_rows[b], group_rows[a])
                )
                if better["score"] != worse["score"]:
                    pairs.append((better, worse))
    rng.shuffle(pairs)
    return pairs


def pairwise_accuracy(rows: list[dict]) -> float:
    pairs = pairs_for(rows, np.random.default_rng(1))
    if not pairs:
        return 1.0
    agree = sum(1 for better, worse in pairs if model.score(better["x"]) > model.score(worse["x"]))
    return agree / len(pairs)


def top1_agreement(rows: list[dict]) -> float:
    group_keys = sorted({row["groupKey"] for row in rows})
    agree = 0
    total = 0
    for group_key in group_keys:
        group_rows = [row for row in rows if row["groupKey"] == group_key]
        if len(group_rows) < 2:
            continue
        best_heuristic = sorted(group_rows, key=lambda row: -row["score"])[0]
        best_model = sorted(group_rows, key=lambda row: -model.score(row["x"]))[0]
        agree += int(best_heuristic["index"] == best_model["index"])
        total += 1
    return agree / total if total else 1.0


def spearman(rows: list[dict]) -> float:
    group_keys = sorted({row["groupKey"] for row in rows})
    rhos = []
    for group_key in group_keys:
        group_rows = [row for row in rows if row["groupKey"] == group_key]
        if len(group_rows) < 3:
            continue
        order = list(range(len(group_rows)))
        by_heuristic = sorted(order, key=lambda position: -group_rows[position]["score"])
        by_model = sorted(order, key=lambda position: -model.score(group_rows[position]["x"]))
        d2 = sum(
            (by_model.index(position) - rank) ** 2
            for rank, position in enumerate(by_heuristic)
        )
        n = len(group_rows)
        rhos.append(1 - (6 * d2) / (n * (n * n - 1)))
    return float(np.mean(rhos)) if rhos else 1.0


dataset = json.loads(DATASET_PATH.read_text(encoding="utf8"))

# Human golden preferences (goal doc Fáze 2 — the ONLY path to "better than
# heuristic"): when scripts/data/intent-ranker-golden.json is reviewed=true,
# golden groups use POSITION-DERIVED labels (best → 1.0 … worst → 0.0)
# instead of the heuristic teacher. Unreviewed/missing file ⇒ heuristic
# teacher everywhere and the model stays in shadow mode.
golden_path = ROOT / "scripts" / "data" / "intent-ranker-golden.json"
golden_orders: dict[str, list[int]] = {}
golden_reviewed = False
if golden_path.exists():
    golden = json.loads(golden_path.read_text(encoding="utf8"))
    golden_reviewed = bool(golden.get("reviewed"))
    if golden_reviewed:
        for combo in golden.get("combos", []):
            golden_orders[combo["groupKey"]] = [int(position) for position in combo["order"]]
        print(f"[golden] reviewed=true — {len(golden_orders)} golden group(s) use human labels")
    else:
        print("[golden] template exists but is NOT reviewed — heuristic teacher stays")

samples: list[dict] = []
for group in dataset["groups"]:
    golden_order = golden_orders.get(group["groupKey"])
    for candidate in group["candidates"]:
        if golden_reviewed and golden_order is not None and candidate["index"] in golden_order:
            position = golden_order.index(candidate["index"])
            denominator = max(1, len(golden_order) - 1)
            label_score = 1.0 - position / denominator
        else:
            label_score = candidate["heuristicScore"]
        samples.append(
            {
                "x": np.array(candidate["features"], dtype=np.float64),
                "score": label_score,
                "index": candidate["index"],
                "groupKey": group["groupKey"],
                "golden": bool(golden_reviewed and golden_order is not None and candidate["index"] in golden_order),
            }
        )
feature_count = len(samples[0]["x"])
rng = make_rng(SEED)
model = MLP([feature_count, *HIDDEN, 1], rng)

group_keys = sorted({sample["groupKey"] for sample in samples})
validation_keys = {group_keys[i] for i in range(len(group_keys)) if i % 5 == 0}
train_rows = [sample for sample in samples if sample["groupKey"] not in validation_keys]
val_rows = [sample for sample in samples if sample["groupKey"] in validation_keys]
print(f"[train] samples total={len(samples)} train={len(train_rows)} val={len(val_rows)} features={feature_count}")

train_pairs = pairs_for(train_rows, rng)
# Full-batch deterministic training: one Adam update per epoch over the
# summed pairwise gradients — per-pair Adam was unstable (oscillated into a
# degenerate constant model on this dataset size).
optimizer_mw = [np.zeros_like(w) for w in model.weights]
optimizer_vw = [np.zeros_like(w) for w in model.weights]
optimizer_mb = [np.zeros_like(b) for b in model.biases]
optimizer_vb = [np.zeros_like(b) for b in model.biases]
beta1, beta2, eps = 0.9, 0.999, 1e-8
for epoch in range(1, EPOCHS + 1):
    rng.shuffle(train_pairs)
    loss_sum = 0.0
    gW = [np.zeros_like(w) for w in model.weights]
    gB = [np.zeros_like(b) for b in model.biases]
    for better, worse in train_pairs:
        loss, pair_gW, pair_gB = model.pair_gradients(better["x"], worse["x"])
        loss_sum += loss
        for index in range(len(gW)):
            gW[index] += pair_gW[index]
            gB[index] += pair_gB[index]
    for index in range(len(model.weights)):
        optimizer_mw[index] = beta1 * optimizer_mw[index] + (1 - beta1) * gW[index] / len(train_pairs)
        optimizer_vw[index] = beta2 * optimizer_vw[index] + (1 - beta2) * (gW[index] / max(1, len(train_pairs))) ** 2
        optimizer_mb[index] = beta1 * optimizer_mb[index] + (1 - beta1) * gB[index] / max(1, len(train_pairs))
        optimizer_vb[index] = beta2 * optimizer_vb[index] + (1 - beta2) * (gB[index] / max(1, len(train_pairs))) ** 2
        mw_hat = optimizer_mw[index] / (1 - beta1**epoch)
        vw_hat = optimizer_vw[index] / (1 - beta2**epoch)
        mb_hat = optimizer_mb[index] / (1 - beta1**epoch)
        vb_hat = optimizer_vb[index] / (1 - beta2**epoch)
        model.weights[index] -= LR * mw_hat / (np.sqrt(vw_hat) + eps)
        model.biases[index] -= LR * mb_hat / (np.sqrt(vb_hat) + eps)
    if epoch % 250 == 0 or epoch == EPOCHS - 1:
        print(f"[train] epoch {epoch} loss={loss_sum / max(1, len(train_pairs)):.5f}")

report = {
    "datasetVersion": dataset["datasetVersion"],
    "featureVersion": "features.v1",
    "featureCount": feature_count,
    "hidden": HIDDEN,
    "epochs": EPOCHS,
    "trainPairs": len(train_pairs),
    "valPairs": len(pairs_for(val_rows, rng)),
    "trainPairwiseAccuracyVsHeuristic": round(pairwise_accuracy(train_rows), 4),
    "valPairwiseAccuracyVsHeuristic": round(pairwise_accuracy(val_rows), 4),
    "valTop1AgreementWithHeuristic": round(top1_agreement(val_rows), 4),
    "valSpearmanVsHeuristic": round(spearman(val_rows), 4),
}
if golden_reviewed:
    golden_rows = [sample for sample in samples if sample.get("golden")]
    report["goldenPairwiseAccuracy"] = round(pairwise_accuracy(golden_rows), 4)
    report["goldenVerdict"] = (
        "ready-for-active"
        if report["goldenPairwiseAccuracy"] >= 0.75
        else "insufficient-golden-fit — stay in shadow and re-curate"
    )
print("[train] report:", json.dumps(report, indent=2))
if golden_reviewed:
    print(f"[train] golden verdict: {report.get('goldenVerdict')}")

MODELS_DIR.mkdir(parents=True, exist_ok=True)
model_bytes = model.export_onnx(feature_count)
model_path = MODELS_DIR / "intent-ranker-v1.onnx"
model_path.write_bytes(model_bytes)
model_hash = hashlib.sha256(model_bytes).hexdigest()
print(f"[train] model written {model_path} ({len(model_bytes) / 1024:.1f} kB) sha256={model_hash[:16]}…")

manifest = {
    "rankerVersion": "ranker.v1",
    "featureVersion": "features.v1",
    "normalizationId": "norm.fixed.v1",
    "featureCount": feature_count,
    "modelPath": "/models/intent-ranker-v1.onnx",
    "inputName": "features",
    "outputName": "score",
    "modelHash": model_hash,
    "hidden": HIDDEN,
    "report": report,
}
(MODELS_DIR / "intent-ranker-v1.manifest.json").write_text(json.dumps(manifest, indent=2), encoding="utf8")

validation_dir = ROOT / "scripts" / "data"
validation_dir.mkdir(parents=True, exist_ok=True)
(validation_dir / "intent-ranker-validation.json").write_text(json.dumps(report, indent=2), encoding="utf8")
print("[train] manifest written")
